### PR TITLE
Feature/type refactor

### PR DESCRIPTION
**Pull request by mabiede**
- typify repository
- refactor structure, improve api
- add possibility to specify transistency
- allow set of effects (and/or lists)
- update backend
- update package dependencies

@aronerben @timohuber fyi

### CODE REVIEW

File: backend/database_pools.ml [hunk 1]
```diff
@@ -1,22 +1,41 @@
 exception Exception of string
 
+let src = Logs.Src.create "guardian.pools"
+
+let find_pool_name : (string * string) list -> string option =
```
[the remark this answers is not in the document]
Comment (mabiede) [reply backend/database_pools.ml]: Not anymore, thanks.

File: backend/mariadb_backend.ml [hunk 2]
```diff
@@ -1,292 +1,411 @@
...
+
+        let revoke_roles ?ctx uuid roles =
+          let open Lwt_result.Syntax in
+          (* TODO: only mark as deleted -> add second revoked_roles column *)
+          let* pre_roles = find_roles ?ctx uuid in
+          let roles' = Roles.diff pre_roles roles in
+          update_roles_request ?ctx uuid roles'
+        ;;
+
+        let find_owner ?ctx id =
```
[the remark this answers is not in the document]
Comment (mabiede) [reply backend/mariadb_backend.ml:287]: renamed

File: lib/guardian.ml [hunk 3]
```diff
@@ -19,211 +34,340 @@ module Util = struct
...
+      : parent
+      =
+      find_opt typ parent_kind
+      |> function
+      | Some parent_fcn -> parent_fcn
+      | None -> default_fcn
+    ;;
+
+    let find_all (typ : TargetRoles.t) : parent list =
+      Map.filter (fun (kind, _) _ -> TargetRoles.equal kind typ) !registered
```
[the remark this answers is not in the document]
Comment (mabiede) [reply lib/guardian.ml]: I updated the "unknown" ones, some are already defined in the modules

**Review by timohuber — APPROVED** [review]

**Review by aronerben — COMMENTED** [review]
Good PR!

**Pull request merged**